packages/tools/pluggable-widgets-tools/configs/rollup-plugin-collect-native-deps.js

```diff
@@ -0,0 +1,123 @@
+    const nativeDependencies = [];
+    const nodeModulesPath = join(outputDir, "node_modules");
+    return {
+        name: "collect-native-deps",
+        async resolveId(source) {
+            if (source.startsWith(".")) {
+                return null;
+            }
+
+            try {
+                const packageFilePath = require.resolve(`${source}/package.json`);
+                const packageDir = dirname(packageFilePath);
+
+                if (await hasNativeCode(packageDir)) {
+                    if (!nativeDependencies.some(x => x.name === source)) {
+                        nativeDependencies.push({ name: source, dir: packageDir });
+                    }
+                    return { id: source, external: true };
+                }
+                return null;
+            } catch (e) {
+                return null;
+            }
+        },
+        async buildStart() {
+            if (shouldCopyNodeModules) {
+                rm("-rf", nodeModulesPath);
+                if (!firstRun) {
+                    nativeDependencies.length = 0;
+                }
+                firstRun = false;
+            }
+        },
+        async writeBundle() {
+            if (shouldCopyNodeModules) {
+                await Promise.all(
+                    nativeDependencies.map(async dependency => {
+                        await copyJsModule(dependency.dir, join(nodeModulesPath, dependency.name));
+                        for (const transitiveDependency of await getTransitiveDependencies(
+                            dependency.name,
+                            externals
+                        )) {
+                            await copyJsModule(
+                                dirname(require.resolve(`${transitiveDependency}/package.json`)),
+                                join(nodeModulesPath, dependency.name, "node_modules", transitiveDependency)
+                            );
+                        }
+                    })
+                );
+                await writeNativeDependenciesJson({ nativeDependencies, outputDir, widgetName });
+            }
+        }
+    };
+}
+
+async function writeNativeDependenciesJson({ nativeDependencies, outputDir, widgetName }) {
+    if (nativeDependencies.length === 0) {
+        return;
+    }
+    const dependencies = {};
+    for (const dependency of nativeDependencies) {
+        dependencies[dependency.name] = (await readJson(join(dependency.dir, "package.json"))).version;
+    }
+    await writeJson(join(outputDir, `${widgetName}.json`), { nativeDependencies: dependencies }, { spaces: 2 });
+}
+
+async function hasNativeCode(dir) {
+    const packageContent = await promises.readdir(dir, { withFileTypes: true });
+
+    if (packageContent.some(file => /^(ios|android|.*\.podspec)$/i.test(file.name))) {
+        return true;
+    }
+
+    for (const file of packageContent) {
+        if (file.isDirectory() && (await hasNativeCode(join(dir, file.name)))) {
```

Comment:
- this test might result in false positives if a package has nested `node_modules`
- can't we find a file by mask then? Like search for "xcode/podspec/gradle that is not inside node_modules" - i think that can be done with a single glob

Reply:
I am not sure, for now I observed few libs containing other libs like react-native-svg which was being correctly copied before.